metaflow/datastore/task_datastore.py

```diff
@@ -204,52 +203,54 @@ def save_artifacts(self, artifacts, force_v4=False):
 
         Parameters
         ----------
-        artifacts : Dict[string -> object]
-            Dictionary containing the human-readable name of the object to save
+        artifacts : Iterator[(string, object)]
+            Iterator over the human-readable name of the object to save
             and the object itself
         force_v4 : boolean or Dict[string -> boolean]
             Indicates whether the artifact should be pickled using the v4
             version of pickle. If a single boolean, applies to all artifacts.
             If a dictionary, applies to the object named only. Defaults to False
             if not present or not specified
+        len_hint: integer
+            Estimated number of items in artifacts_iter
         """
         artifact_names = []
-        to_save = []
-        originals = []
-        for name, obj in artifacts.items():
-            original_type = str(type(obj))
-            do_v4 = force_v4 and \
-                force_v4 if isinstance(force_v4, bool) else \
-                    force_v4.get('name', False)
-            if do_v4:
-                encode_type = 'pickle-v4'
-                if encode_type not in self._encodings:
-                    raise DataException(
-                        "Artifact *%s* requires a serialization encoding that "
-                        "requires Python 3.4 or newer." % name)
-                obj = pickle.dumps(obj, protocol=4)
-            else:
-                try:
-                    obj = pickle.dumps(obj, protocol=2)
-                    encode_type = 'pickle-v2'
-                except (SystemError, OverflowError):
-                    encode_type = 'pickle-v4'
+
+        def pickle_iter():
+            for name, obj in artifacts_iter:
+                original_type = str(type(obj))
```

Comment:
Nit: you actually don't need original_type anymore.

Reply:
good point!